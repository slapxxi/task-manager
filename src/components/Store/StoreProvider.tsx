import { toEpochTime } from '@lib';
import fetchDatabase from '@local/services/fetchDatabase';
import saveDatabase from '@local/services/saveDatabase';
import {
  Database,
  EpochTime,
  ID,
  Project,
  StoreEntry,
  StoreProject,
  StoreTag,
  StoreTask,
  Tag,
  Task,
} from '@local/types';
import { cloneDeep } from 'lodash-es';
import find from 'lodash-es/find';
import flatten from 'lodash-es/flatten';
import includes from 'lodash-es/includes';
import map from 'lodash-es/map';
import reduce from 'lodash-es/reduce';
import uniq from 'lodash-es/uniq';
import React, { useEffect, useReducer } from 'react';
import { Provider } from './context';

enum ActionType {
  receiveData = 'RECEIVE_DATA',
  requestData = 'REQUEST_DATA',
  updateTask = 'UPDATE_TASK',
  deleteTask = 'DELETE_TASK',
  updateProject = 'UPDATE_PROJECT',
  deleteProject = 'DELETE_PROJECT',
}

type StoreAction =
  | { type: ActionType.requestData }
  | { type: ActionType.receiveData; payload: Database }
  | { type: ActionType.updateTask; payload: Task }
  | { type: ActionType.deleteTask; payload: Task }
  | { type: ActionType.updateProject; payload: Project }
  | { type: ActionType.deleteProject; payload: Project };

interface State {
  readonly tasks: StoreEntry<StoreTask>;
  readonly tags: StoreEntry<StoreTag>;
  readonly projects: StoreEntry<StoreProject>;
  readonly isLoading: boolean;
  readonly lastUpdated: EpochTime;
}

interface Props {
  initialValue?: State;
  children: React.ReactNode;
}

const defaultState: State = {
  tags: {},
  tasks: {},
  projects: {},
  isLoading: false,
  lastUpdated: 0,
};

function StoreProvider({ initialValue = defaultState, children }: Props) {
  const [state, dispatch] = useReducer<State, StoreAction>(
    storeReducer,
    cloneDeep(initialValue),
  );

  useEffect(() => {
    dispatch({ type: ActionType.requestData });
    fetchDatabase((db) => dispatch({ type: ActionType.receiveData, payload: db }));
  }, []);

  useEffect(
    () => {
      if (state.lastUpdated === 0) {
        return;
      }
      saveDatabase({ tasks: state.tasks, tags: state.tags, projects: state.projects });
    },
    [state.lastUpdated],
  );

  return <Provider value={{ ...state, dispatch }}>{children}</Provider>;
}

function storeReducer(state: State, action: StoreAction): State {
  switch (action.type) {
    case ActionType.requestData:
      return { ...state, isLoading: true };
    case ActionType.receiveData:
      return { ...action.payload, isLoading: false, lastUpdated: state.lastUpdated };
    case ActionType.updateTask:
      return updateTask(state, action.payload);
    case ActionType.deleteTask:
      return deleteTask(state, action.payload);
    case ActionType.updateProject:
      return updateProject(state, action.payload);
    case ActionType.deleteProject:
      return deleteProject(state, action.payload);
    default:
      return state;
  }
}

function updateTask(state: State, task: Task): State {
  const { tags: updatedTags = {} } = task.tags.reduce(
    (updatedState, tag) => updateTag(updatedState, tag),
    state,
  );
  return {
    ...state,
    tags: updatedTags,
    tasks: {
      ...state.tasks,
      [task.id]: normalizeTask({
        ...task,
        tags: task.tags.map((t) => findMatchingTag(state, t)),
      }),
    },
    lastUpdated: Date.now(),
  };
}

function updateTag(state: State, tag: Tag): State {
  const updatedTag = findMatchingTag(state, tag);
  return {
    ...state,
    tags: { ...state.tags, [updatedTag.id]: { name: updatedTag.name } },
  };
}

function findMatchingTag(state: State, tag: Tag): Tag {
  const matchingTag = find(
    map(state.tags, (t, id) => ({ id, name: t.name })),
    (t) =>
      t.name.toLowerCase().trim() === tag.name.toLowerCase().trim() || t.id === tag.id,
  );
  if (matchingTag) {
    return matchingTag;
  } else {
    return tag;
  }
}

function normalizeTask(task: Task): StoreTask {
  const dbTask = {
    ...task,
    description: task.description || null,
    project: task.project || null,
    tags: (task.tags as Tag[]).map((t) => t.id),
    deadline: task.deadline ? toEpochTime(task.deadline) : null,
    createdAt: toEpochTime(task.createdAt),
  };
  delete dbTask.id;
  return dbTask;
}

function updateProject(state: State, project: Project): State {
  return {
    ...state,
    projects: { ...state.projects, [project.id]: normalizeProject(project) },
    lastUpdated: Date.now(),
  };
}

function normalizeProject(project: Project): StoreProject {
  return { name: project.name };
}

function deleteTask(state: State, task: Task): State {
  const updatedState = { ...state };
  delete updatedState.tasks[task.id];
  return deleteExtranousTags(updatedState);
}

function deleteExtranousTags(state: State): State {
  const updatedTags = reduce(
    state.tags,
    (acc, tag, tagId) => {
      if (isUsedAnywhere(state, tagId)) {
        return { ...acc, [tagId]: tag };
      }
      return acc;
    },
    {},
  );
  return { ...state, tags: updatedTags };
}

function isUsedAnywhere(state: State, tagId: ID): boolean {
  const tagsUsedElsewhere = getTagsUsedInTasks(state);
  if (includes(tagsUsedElsewhere, tagId)) {
    return true;
  }
  return false;
}

function getTagsUsedInTasks(state: State): ID[] {
  return uniq(flatten(map(state.tasks, (task) => task.tags)));
}

function deleteProject(state: State, project: Project): State {
  const updatedState = { ...state, tasks: state.tasks };
  delete updatedState.projects[project.id];
  return removeProjectFromTasks(state, project);
}

function removeProjectFromTasks(state: State, project: Project): State {
  const updatedTasks = reduce(
    state.tasks,
    (acc, task, taskId) => {
      if (task.project === project.id) {
        return { ...acc, [taskId]: { ...task, project: '' } };
      }
      return { ...acc, [taskId]: task };
    },
    {},
  );
  return { ...state, tasks: updatedTasks };
}

export { ActionType, State, StoreAction };

export default StoreProvider;
