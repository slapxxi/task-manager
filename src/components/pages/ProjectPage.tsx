import { Project, Store } from '@local/components';
import * as React from 'react';

interface Props {
  projectID: ID;
}

function ProjectPage({ projectID }: Props) {
  return (
    <div>
      <Store>
        {({ projects, actions }) => (
          <Project
            project={projects.filter((p) => p.id === projectID)[0]}
            onEdit={actions.updateProject}
            onEditTask={actions.updateTask}
            onDeleteTask={actions.deleteTask}
          />
        )}
      </Store>
    </div>
  );
}

export default ProjectPage;
