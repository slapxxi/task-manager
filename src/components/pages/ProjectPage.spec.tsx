import StoreProvider from '@local/components/Store/StoreProvider';
import React from 'react';
import { MemoryRouter as Router, Route } from 'react-router-dom';
import { fireEvent, render } from 'react-testing-library';
import ProjectPage from './ProjectPage';

jest.mock('@local/components/IconSystem/IconSystem');
jest.mock('@local/hooks/useWorker');

const mockStore = {
  tasks: {
    'test-task': {
      title: 'Test Task',
      project: 'test-project',
      completed: false,
      createdAt: 0,
      tags: [],
      subtasks: [],
    },
  },
  tags: {},
  projects: { 'test-project': { name: 'Test Project' } },
  isLoading: false,
  isSyncing: false,
  lastUpdated: 0,
};

it('redirects to home page if there is no matching project', () => {
  const { container } = render(
    <StoreProvider>
      <Router initialEntries={['/projects/test-project']}>
        <div>
          <Route path="/" render={() => <div>Home</div>} exact />
          <Route
            path="/projects/:id"
            render={({ match }) => <ProjectPage projectID={match.params.id} />}
          />
        </div>
      </Router>
      ,
    </StoreProvider>,
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders project', () => {
  const { container } = render(
    <StoreProvider initialValue={mockStore}>
      <Router initialEntries={['/projects/test-project']}>
        <Route
          path="/projects/:id"
          render={({ match }) => <ProjectPage projectID={match.params.id} />}
        />
      </Router>
      ,
    </StoreProvider>,
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('allows to change project title', () => {
  const { getByTestId } = render(
    <StoreProvider initialValue={mockStore}>
      <Router initialEntries={['/projects/test-project']}>
        <Route
          path="/projects/:id"
          render={({ match }) => <ProjectPage projectID={match.params.id} />}
        />
      </Router>
      ,
    </StoreProvider>,
  );
  const projectName = getByTestId('project-name') as HTMLTextAreaElement;

  fireEvent.change(projectName, { target: { value: 'Changed Title' } });

  expect(projectName.value).toEqual('Changed Title');
});
