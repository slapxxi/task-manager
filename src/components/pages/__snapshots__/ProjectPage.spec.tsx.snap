// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redirects to home page if there is no matching project 1`] = `
<div>
  <div>
    Home
  </div>
</div>
`;

exports[`renders project 1`] = `
<div
  class="projectPage"
>
  <div
    class="toolbar"
  >
    <ul
      class="container"
    >
      <li
        class="item"
      >
        <a
          class="link"
          href="/"
        >
          <div
            name="inbox"
            size="20"
          >
            icon-
            inbox
          </div>
        </a>
      </li>
      <li
        class="item"
      >
        <div
          class="icon"
          name="arrow-right"
          size="10"
        >
          icon-
          arrow-right
        </div>
        <a
          class="link"
          href="/projects"
        >
          projects
        </a>
      </li>
      <li
        class="item"
      >
        <div
          class="icon"
          name="arrow-right"
          size="10"
        >
          icon-
          arrow-right
        </div>
        <a
          aria-current="page"
          class="link active"
          href="/projects/test-project"
        >
          test-project
        </a>
      </li>
    </ul>
    <div>
      Order 
      <div
        name="arrow-down"
        size="10"
      >
        icon-
        arrow-down
      </div>
    </div>
    <div>
      Filter 
      <div
        name="arrow-down"
        size="10"
      >
        icon-
        arrow-down
      </div>
    </div>
  </div>
  <header
    class="header"
  >
    <div>
      <div
        class="container"
        style="height: 22px;"
      >
        <input
          class="input"
          type="checkbox"
        />
        <svg
          height="22"
          viewBox="0 0 100 100"
          width="22"
        >
          <circle
            cx="50"
            cy="50"
            fill="none"
            r="44"
            stroke="#99a"
            stroke-width="8"
          />
          <path
            d="M 50 17 A 33 33 0 0 0 50 17 L 50 50 L 50 17 Z"
            fill="#aab"
            stroke="none"
            stroke-width="6"
          />
          /&gt;
        </svg>
      </div>
    </div>
    <textarea
      class="editableTitle sc-htpNat gyAANP"
      data-testid="project-name"
      placeholder="Project Name..."
      rows="1"
      style="overflow-x: hidden; word-wrap: break-word; overflow-y: hidden;"
    >
      Test Project
    </textarea>
    <button
      class="menu sc-bdVaJa fFWIpk"
    >
      <div
        name="dots"
        size="22"
      >
        icon-
        dots
      </div>
    </button>
  </header>
  <div
    class="subtitle"
  >
    1
     Tasks
  </div>
  <div
    class="content"
  >
    <ul
      class="tasks"
    >
      <li>
        <div
          class="task sc-bxivhb iaqica"
        >
          <header
            class="taskHeader"
          >
            <div>
              <div
                style="position: relative; display: flex; align-items: center; height: 22px;"
              >
                <input
                  class="input"
                  data-testid="checkbox"
                  type="checkbox"
                />
                <svg
                  clip-rule="evenodd"
                  height="22"
                  preserveAspectRatio="none"
                  viewBox="0 0 100 100"
                  width="22"
                >
                  <circle
                    cx="50"
                    cy="50"
                    fill="none"
                    r="44"
                    stroke="#99a"
                    stroke-width="8"
                  />
                </svg>
              </div>
            </div>
            <textarea
              class="taskTitle sc-ifAKCX iMtRMp sc-htpNat gyAANP"
              data-testid="title"
              placeholder="Title..."
              rows="1"
              style="overflow-x: hidden; word-wrap: break-word; overflow-y: hidden;"
            >
              Test Task
            </textarea>
            <button
              class="todayButton sc-bdVaJa fFWIpk"
              data-testid="setToday"
            >
              <div
                name="star"
                size="18"
              >
                icon-
                star
              </div>
            </button>
            <button
              class="detailsButton sc-bdVaJa fFWIpk"
              data-testid="expand"
            >
              <div
                name="arrow-down"
                size="16"
              >
                icon-
                arrow-down
              </div>
            </button>
          </header>
          <div
            style="overflow: hidden; height: 0px; opacity: 0;"
          >
            <div
              class="sc-EHOje ddexpk"
            >
              <textarea
                class="taskDescription sc-htpNat gyAANP"
                data-testid="description"
                placeholder="Notes..."
                rows="1"
                style="overflow-x: hidden; word-wrap: break-word; overflow-y: hidden;"
              />
              <footer
                class="taskFooter"
              >
                <div
                  class="footerTags"
                >
                  <ul
                    class="tags"
                    data-testid="tags"
                  />
                  <input
                    class="input"
                    data-testid="input"
                    placeholder="Tag..."
                    type="text"
                  />
                </div>
                <div
                  class="footerIconDangerous"
                  data-testid="delete"
                  name="trashbin"
                  size="16"
                >
                  icon-
                  trashbin
                </div>
                <div
                  class="footerIcon"
                  name="list"
                  size="16"
                >
                  icon-
                  list
                </div>
                <div
                  class="footerIcon"
                  name="flag"
                  size="16"
                >
                  icon-
                  flag
                </div>
              </footer>
            </div>
          </div>
        </div>
      </li>
    </ul>
    <footer
      class="footer"
    >
      <button
        class="button sc-bdVaJa fFWIpk"
      >
        Create Task
      </button>
    </footer>
  </div>
</div>
`;
